File: ashgf/algorithms/sges.py
"""Self-Guided Evolution Strategies (SGES) optimizer."""

from __future__ import annotations

import logging
from typing import Callable

import numpy as np

from ashgf.algorithms.base import BaseOptimizer
from ashgf.gradient.estimators import _parallel_eval, gaussian_smoothing
from ashgf.gradient.sampling import compute_directions, compute_directions_sges

logger = logging.getLogger(__name__)

__all__ = ["SGES"]


class SGES(BaseOptimizer):
    """Self-Guided Evolution Strategies.

    Extends Gaussian smoothing by adaptively mixing random directions
    with directions sampled from the gradient-history subspace.

    .. note::

       **Semantica di ``alpha``.**  In questa implementazione ``alpha``
       rappresenta la probabilità di campionare una direzione **casuale**
       (isotropica).  Il valore ``1 - alpha`` è quindi la probabilità di
       campionare dal sottospazio dei gradienti storici.  Questa scelta è
       coerente con ``compute_directions_sges``, dove
       ``choices = binomial(dim, 1 - alpha)``.

       Nella tesi, il parametro :math:`\\alpha` ha il significato opposto
       (probabilità di campionare dal sottospazio gradiente).  Di
       conseguenza, la regola di aggiornamento è **invertita** rispetto
       alla formulazione della tesi: quando il sottospazio gradiente è
       migliore (:math:`r < \\hat{r}`), ``alpha`` viene **diminuita** per
       favorire lo sfruttamento.

    Parameters
    ----------
    lr : float
        Fixed learning rate.
    sigma : float
        Smoothing bandwidth.
    k : int
        Parameter controlling buffer management.
    k1 : float
        Upper bound for alpha.
    k2 : float
        Lower bound for alpha.
    alpha : float
        Initial probability of sampling a **random** (isotropic) direction.
        ``1 - alpha`` is the probability of sampling from the gradient
        subspace.
    delta : float
        Multiplicative factor for alpha update.
    t : int
        Number of pure-random warm-up iterations.
    seed : int
        Random seed.
    eps : float
        Convergence threshold.
    """

    kind = "SGES"

    def __init__(
        self,
        lr: float = 1e-4,
        sigma: float = 1e-4,
        k: int = 50,
        k1: float = 0.9,
        k2: float = 0.1,
        alpha: float = 0.5,
        delta: float = 1.1,
        t: int = 50,
        seed: int = 2003,
        eps: float = 1e-8,
    ) -> None:
        super().__init__(seed=seed, eps=eps)
        if lr <= 0:
            raise ValueError(f"Learning rate must be > 0, got {lr}")
        if sigma <= 0:
            raise ValueError(f"sigma must be > 0, got {sigma}")

        self.lr = lr
        self.sigma = sigma
        self.k = k
        self.k1 = k1
        self.k2 = k2
        self.alpha = alpha
        self.delta = delta
        self.t = t

        # Internal state — circular buffer for gradient history
        self._G_buffer: np.ndarray | None = None  # (t, dim) pre-allocated
        self._G_count: int = 0  # number of gradients stored so far
        self._G_idx: int = 0  # next write position
        self._current_alpha = alpha

    def _get_step_size(self) -> float:
        return self.lr

    def _setup(self, f: Callable[[np.ndarray], float], dim: int, x: np.ndarray) -> None:
        self._G_buffer = np.zeros((self.t, dim))
        self._G_count = 0
        self._G_idx = 0
        self._current_alpha = self.alpha

    def _post_iteration(
        self, iteration: int, x: np.ndarray, grad: np.ndarray, f_val: float
    ) -> None:
        """Update alpha based on relative performance of gradient vs random directions.

        .. note::

           **Corretto (bug #1).**  Nella tesi, quando
           :math:`\\hat{r}_{\\mathbf{G}} < \\hat{r}_{\\mathbf{G}}^{\\perp}`
           (il sottospazio gradiente produce minimi migliori), :math:`\\alpha`
           viene **aumentato** per favorire lo sfruttamento.  In questo codice
           ``alpha`` è la probabilità di direzioni **casuali**, quindi quando
           ``r < r_hat`` ``alpha`` viene **diminuita** (effetto equivalente:
           più direzioni dal gradiente).
        """
        if iteration < self.t:
            return

        # The evaluations were stored by the gradient estimator
        if not hasattr(self, "_last_evaluations") or not hasattr(self, "_last_M"):
            return

        evaluations = self._last_evaluations  # (2*dim,) array
        M = self._last_M
        dim = len(x)

        if M <= 0 or M >= dim:
            return

        # Vectorised: reshape to (dim, 2) and take min per row
        evals_pairs = evaluations.reshape(-1, 2)  # (dim, 2)
        min_per_dir = np.min(evals_pairs, axis=1)  # (dim,)

        # r = mean of minima over gradient-subspace directions (0:M)
        r = float(np.mean(min_per_dir[:M]))
        # r_hat = mean of minima over random directions (M:dim)
        r_hat = float(np.mean(min_per_dir[M:]))

        # FIXED (bug #1): alpha is probability of RANDOM directions,
        # so when gradient subspace is better (r < r_hat) we DECREASE alpha
        # (favour gradient subspace).  When random is better, INCREASE alpha.
        if r < r_hat:
            # Gradient subspace is more promising → decrease alpha (less random)
            self._current_alpha = max(self._current_alpha / self.delta, self.k2)
        else:
            # Random subspace is more promising (or tie) → increase alpha (more exploration)
            self._current_alpha = min(self.delta * self._current_alpha, self.k1)

    def grad_estimator(
        self, x: np.ndarray, f: Callable[[np.ndarray], float]
    ) -> np.ndarray:
        """Estimate gradient using SGES direction sampling.

        FIXED: No longer seeds RNG inside this method (bug 1.5.3).
        FIXED: Correctly uses sges directions when buffer is available (bug 1.5.1).
        """
        dim = len(x)

        # Use SGES directions once we have at least t-1 gradients in the buffer
        # (matching original code behavior: at iteration t, G has t-1 entries)
        if self._G_count >= self.t - 1:
            # Use SGES directions mixing gradients + random
            directions, M = compute_directions_sges(
                dim, self._G_buffer[: self._G_count], self._current_alpha
            )
            self._last_M = M

            # Collect evaluations for alpha update
            sigma_dirs = self.sigma * directions  # (dim, dim) pre-scaled
            points: list[np.ndarray] = []
            for i in range(dim):
                d = sigma_dirs[i]
                points.append(x + d)
                points.append(x - d)

            # Evaluate (parallel if ASHGF_N_JOBS > 1)
            results = _parallel_eval(f, points)
            evaluations = np.asarray(results)

            # ---- Vectorised gradient assembly ----
            # diff[i] = f(x + σ·d_i) - f(x - σ·d_i)
            diff = evaluations[0::2] - evaluations[1::2]  # (dim,)
            grad = np.dot(diff, directions) / (2.0 * self.sigma * dim)

            self._last_evaluations = evaluations
        else:
            # Pure random directions during warm-up
            directions = compute_directions(dim)
            grad = gaussian_smoothing(x, f, self.sigma, directions)
            self._last_M = 0
            self._last_evaluations = np.array([])

        # Append to circular gradient buffer
        self._G_buffer[self._G_idx] = grad
        self._G_idx = (self._G_idx + 1) % self.t
        self._G_count = min(self._G_count + 1, self.t)

        return grad

File: src/algorithms/sges.rs
//! Self-Guided Evolution Strategies (SGES) optimiser.

use ndarray::{Array1, Array2};

use crate::algorithms::base::Optimizer;
use crate::gradient::{compute_directions, compute_directions_sges, gaussian_smoothing};
use crate::utils::{parallel_eval, SeededRng};

/// Self-Guided Evolution Strategies.
///
/// Extends Gaussian smoothing by adaptively mixing random directions
/// with directions sampled from the gradient-history subspace.
///
/// .. note::
///
///    **Alpha semantics.**  `alpha` is the probability of sampling a
///    **random** (isotropic) direction.  `1 - alpha` is the probability
///    of sampling from the gradient subspace.  This is inverted relative
///    to the thesis, where α denotes gradient-subspace probability.
///    The update rule is correspondingly inverted.
pub struct SGES {
    /// Fixed learning rate.
    pub lr: f64,
    /// Smoothing bandwidth.
    pub sigma: f64,
    /// Upper bound for alpha.
    pub k1: f64,
    /// Lower bound for alpha.
    pub k2: f64,
    /// Initial alpha (prob. of random direction).
    pub alpha_init: f64,
    /// Multiplicative factor for alpha update.
    pub delta: f64,
    /// Number of pure-random warm-up iterations.
    pub t: usize,
    /// Number of parallel threads for function evaluation.
    pub n_jobs: usize,
    eps: f64,

    // -- Adaptive state --
    /// Current alpha.
    current_alpha: f64,
    /// Gradient buffer: (t, dim).
    g_buffer: Option<Array2<f64>>,
    g_count: usize,
    g_idx: usize,
    /// Last evaluations for alpha update.
    last_evaluations: Option<Array1<f64>>,
    last_m: usize,
}

impl SGES {
    pub fn new(
        lr: f64,
        sigma: f64,
        k1: f64,
        k2: f64,
        alpha: f64,
        delta: f64,
        t: usize,
        eps: f64,
    ) -> Self {
        assert!(lr > 0.0);
        assert!(sigma > 0.0);
        assert!(k2 <= k1);
        assert!((0.0..=1.0).contains(&alpha));
        Self {
            lr,
            sigma,
            k1,
            k2,
            alpha_init: alpha,
            delta,
            t,
            eps,
            n_jobs: 0,
            current_alpha: alpha,
            g_buffer: None,
            g_count: 0,
            g_idx: 0,
            last_evaluations: None,
            last_m: 0,
        }
    }
}

impl Optimizer for SGES {
    fn kind(&self) -> &'static str {
        "SGES"
    }

    fn eps(&self) -> f64 {
        self.eps
    }

    fn step_size(&self) -> f64 {
        self.lr
    }

    fn setup(&mut self, _f: &(dyn Fn(&Array1<f64>) -> f64 + Sync), dim: usize, _x: &Array1<f64>) {
        self.g_buffer = Some(Array2::zeros((self.t, dim)));
        self.g_count = 0;
        self.g_idx = 0;
        self.current_alpha = self.alpha_init;
    }

    fn post_iteration(
        &mut self,
        iteration: usize,
        _x: &Array1<f64>,
        _grad: &Array1<f64>,
        _f_val: f64,
    ) {
        if iteration < self.t {
            return;
        }

        let evaluations = match &self.last_evaluations {
            Some(ev) => ev,
            None => return,
        };
        let m_val = self.last_m;

        if m_val == 0 || m_val >= evaluations.len() / 2 {
            return;
        }

        // Reshape to (dim, 2) and take min per pair
        let dim = evaluations.len() / 2;
        let mut min_per_dir = Array1::zeros(dim);
        for i in 0..dim {
            let a = evaluations[2 * i];
            let b = evaluations[2 * i + 1];
            min_per_dir[i] = a.min(b);
        }

        let r = min_per_dir.slice(ndarray::s![..m_val]).mean().unwrap();
        let r_hat = min_per_dir.slice(ndarray::s![m_val..]).mean().unwrap();

        // Inverted logic: alpha = probability of RANDOM direction
        if r < r_hat {
            // Gradient subspace better → decrease alpha (less random)
            self.current_alpha = (self.current_alpha / self.delta).max(self.k2);
        } else {
            // Random better → increase alpha (more exploration)
            self.current_alpha = (self.delta * self.current_alpha).min(self.k1);
        }
    }

    fn grad_estimator(
        &mut self,
        x: &Array1<f64>,
        f: &(dyn Fn(&Array1<f64>) -> f64 + Sync),
        rng: &mut SeededRng,
    ) -> Array1<f64> {
        let dim = x.len();
        let n_jobs = if self.n_jobs > 0 {
            self.n_jobs
        } else {
            rayon::current_num_threads()
        };

        let grad = if self.g_count >= self.t - 1 {
            // Use SGES directions
            let g = self.g_buffer.as_ref().unwrap();
            let g_slice = g.slice(ndarray::s![..self.g_count, ..]);
            let (directions, m_val) =
                compute_directions_sges(dim, &g_slice.to_owned(), self.current_alpha, rng);
            self.last_m = m_val;

            // Collect evaluations (parallel)
            let sigma_dirs = self.sigma * &directions;
            let mut points = Vec::with_capacity(2 * dim);
            for i in 0..dim {
                let d = sigma_dirs.row(i).to_owned();
                points.push(x + &d);
                points.push(x - &d);
            }
            let results = parallel_eval(f, &points, n_jobs);
            self.last_evaluations = Some(Array1::from_vec(results.clone()));

            // Gradient assembly
            let results_arr = Array1::from_vec(results);
            let diff = results_arr.slice(ndarray::s![0..;2]).to_owned()
                - &results_arr.slice(ndarray::s![1..;2]);
            diff.dot(&directions) / (2.0 * self.sigma * dim as f64)
        } else {
            // Warm-up: pure random
            let directions = compute_directions(dim, rng);
            self.last_m = 0;
            self.last_evaluations = Some(Array1::zeros(2 * dim));
            gaussian_smoothing(x, f, self.sigma, &directions, n_jobs)
        };

        // Update gradient buffer
        if let Some(ref mut buf) = self.g_buffer {
            buf.row_mut(self.g_idx).assign(&grad);
            self.g_idx = (self.g_idx + 1) % self.t;
            self.g_count = (self.g_count + 1).min(self.t);
        }

        grad
    }
}

impl Default for SGES {
    fn default() -> Self {
        Self::new(1e-4, 1e-4, 0.9, 0.1, 0.5, 1.1, 50, 1e-8)
    }
}
